rewrite/rewrite/python/format/tabs_and_indents_visitor.py

```diff
@@ -127,6 +127,24 @@ def visit_space(self, space: Optional[Space], loc: Optional[Union[PySpace.Locati
 
         return s
 
+    @staticmethod
+    def compute_first_parameter_offset(method: j.MethodDeclaration, first_arg: J, cursor: Cursor) -> int:
+        class FirstArgPrinter(PrintOutputCapture[TreeVisitor]):
+            def append(self, text: Optional[str] = None) -> PrintOutputCapture[P]:
+                if self._context.cursor.value is first_arg:
+                    raise RecipeRunException()
+                return super().append(text)
+
+        printer = method.printer(cursor)
+        capture = FirstArgPrinter(printer)
+        try:
+            printer.visit(method, capture, cursor.parent_or_throw)
+        except RecipeRunException:
+            source = capture.get_out()
+            def_idx = source.index("def")
+            async_idx = source.find("async")
+            start_idx = async_idx if async_idx != -1 and async_idx < def_idx else def_idx
```

Comment:
I am a little bit worried that this could go wrong if there is a decorator or comment including the substring `def` or `async` somewhere. It would be safer if we in this `FirstArgPrinter` would detect when either of these keywords is written and store that in a variable.

Reply:
I am also wondering if this works correctly when for example a decorator is on the same line as the `def` keyword as in:
```py
@decorator def my_function(a,
        b):
    pass
```
I suppose this would be easy to test.

Reply:
That is not valid Python